zerver/tests/test_push_notifications.py

```diff
@@ -139,6 +146,186 @@ def get_generic_payload(self, method: str = "register") -> Dict[str, Any]:
+
+        endpoint = "/api/v1/mobile_push/test_notification"
+        time_now = now()
+
+        # 1. First test for an android device.
+        # 2. Then test for an apple device.
+        # 3. Then test without submitting a specific token,
+        #    meaning both devices should get notified.
+
+        with mock.patch(
+            "zerver.lib.push_notifications.send_android_push_notification"
+        ) as mock_send_android_push_notification, time_machine.travel(time_now, tick=False):
+            result = self.api_post(user, endpoint, {"token": android_token}, subdomain="zulip")
+
+        expected_android_payload = {
+            "server": "testserver",
+            "realm_id": user.realm_id,
+            "realm_uri": "http://zulip.testserver",
+            "user_id": user.id,
+            "event": "test-by-device-token",
+            "time": datetime_to_timestamp(time_now),
+        }
+        expected_gcm_options = {"priority": "high"}
+        mock_send_android_push_notification.assert_called_once_with(
+            UserPushIdentityCompat(user_id=user.id, user_uuid=str(user.uuid)),
+            [android_device],
+            expected_android_payload,
+            expected_gcm_options,
+            remote=None,
+        )
+        self.assert_json_success(result)
+
+        with mock.patch(
+            "zerver.lib.push_notifications.send_apple_push_notification"
+        ) as mock_send_apple_push_notification, time_machine.travel(time_now, tick=False):
+            result = self.api_post(user, endpoint, {"token": apple_token}, subdomain="zulip")
+
+        expected_apple_payload = {
+            "alert": {
+                "title": "Test notification",
+                "body": "This is a test notification from http://zulip.testserver.",
+            },
+            "sound": "default",
+            "custom": {
+                "zulip": {
+                    "server": "testserver",
+                    "realm_id": user.realm_id,
+                    "realm_uri": "http://zulip.testserver",
+                    "user_id": user.id,
+                    "event": "test-by-device-token",
+                }
+            },
+        }
+        mock_send_apple_push_notification.assert_called_once_with(
+            UserPushIdentityCompat(user_id=user.id, user_uuid=str(user.uuid)),
+            [apple_device],
+            expected_apple_payload,
+            remote=None,
+        )
+        self.assert_json_success(result)
+
+        # Test without submitting a token value. Both devices should get notified.
+        with mock.patch(
+            "zerver.lib.push_notifications.send_apple_push_notification"
+        ) as mock_send_apple_push_notification, mock.patch(
+            "zerver.lib.push_notifications.send_android_push_notification"
+        ) as mock_send_android_push_notification, time_machine.travel(
+            time_now, tick=False
+        ):
+            result = self.api_post(user, endpoint, {"token": ""}, subdomain="zulip")
```

Comment:
Yeah, this is an explicit `""` string as a token, rather than just not passing the parameter; makes me think our `REQ` configuration is off.